handler/src/main/java/io/netty/handler/ssl/SslContext.java

```diff
@@ -796,4 +805,54 @@ protected static PKCS8EncodedKeySpec generateKeySpec(char[] password, byte[] key
 
         return encryptedPrivateKeyInfo.getKeySpec(cipher);
     }
+
+    /**
+     * Generates a new {@link KeyStore}.
+     *
+     * @param certChainFile a X.509 certificate chain file in PEM format,
+     * @param keyFile a PKCS#8 private key file in PEM format,
+     * @param keyPasswordChars the password of the {@code keyFile}.
+     *                    {@code null} if it's not password-protected.
+     * @return generated {@link KeyStore}.
+     */
+    protected static KeyStore buildKeyStore(File certChainFile, File keyFile, char[] keyPasswordChars)
+            throws KeyStoreException, NoSuchAlgorithmException,
+                   NoSuchPaddingException, InvalidKeySpecException, InvalidAlgorithmParameterException,
+                   CertificateException, KeyException, IOException {
+        ByteBuf encodedKeyBuf = PemReader.readPrivateKey(keyFile);
+        byte[] encodedKey = new byte[encodedKeyBuf.readableBytes()];
+        encodedKeyBuf.readBytes(encodedKey).release();
+
+        PKCS8EncodedKeySpec encodedKeySpec = generateKeySpec(keyPasswordChars, encodedKey);
+
+        PrivateKey key;
+        try {
+            key = KeyFactory.getInstance("RSA").generatePrivate(encodedKeySpec);
+        } catch (InvalidKeySpecException ignore) {
+            try {
+                key = KeyFactory.getInstance("DSA").generatePrivate(encodedKeySpec);
+            } catch (InvalidKeySpecException ignore2) {
+                key = KeyFactory.getInstance("EC").generatePrivate(encodedKeySpec);
```

Comment:
Perhaps it's better to throw a new `InvalidKeySpecException` with a proper message? (i.e. Neither RSA, DSA nor EC worked)